docs/client-design.md

```diff
@@ -0,0 +1,306 @@
+	 */
+	enum ConnectionState get_state() const;
+	
+	/** 
+	 * If the connection is an erroneous state (see `status`), return the 
+	 * connection error that caused it. The same error is also passed to 
+	 * response callbacks, and the same error is returned by 
+	 * `Request::get_error`. 
+	 */
+	std::optional<ConnectionError> &get_error() & const;
+	
+	/* An abstract request's interface. A request object is always returned. */
+	Request some_request(/* options */);
+};
+```
+
+#### Connection State
+
+```c++
+enum ConnectionState {
+	CONNECTION_INITIAL = 0,
+	CONNECTION_AUTH = 1,
+	CONNECTION_ACTIVE = 2,
+	CONNECTION_ERROR = 3,
+	CONNECTION_ERROR_RECONNECT = 4,
+};
+```
+
+#### Connect Options
+
+```c++
+/* Extend `ConnectionOptions` with an option for the reconnection feature. */
+struct ConnectOptions {
+	/* All existing options. */
+	
+	/** 
+	 * In the event of a broken connection, the interval in which the stream 
+	 * tries to re-establish the connection. 
+	 */
+	static constexpr size_t DEFAULT_RECONNECTION_INTERVAL = 2;
+	size_t reconnection_interval = DEFAULT_RECONNECTION_INTERVAL;
+};
+```
+
+### Requests
+
+```c++
+/** Encapsulates management of a request issued through a connection. */
+class Request {
+public:
+	/** 
+	 * Callback called when response is ready. Since the callback has a fixed 
+	 * signature, we need to allow for capturing additional context using 
+	 * lambdas. Hence, we use `std::function` for type erasure.
+	 * 
+	 * See `Connection::get_error` for details about the `error` parameter.
+	 */ 
+	using request_cb_f = 
+		std::function<void(Request &&request, Connection &connection)>;
+	
+	/** Set a callback called when the response is ready. */
+	void set_callback(request_cb_f request_cb) &&;
+	
+	/** Return the request's status. */
+	enum RequestStatus get_status();
+	
+	/**
+	 * Return a connection error, if any. See `Connection::get_error` for 
+	 * details.
+	 */
+	std::optional<ConnectionError> &get_error() & const;
+	
+	/**
+	 * Cancel the request, ending the lifetime of the corresponding response.
+	 */
+	void cancel();
+	
+	/** 
+	 * Return the response, if any. A response is available, iff:
+	 * 1. The response has been received by the connection.
+	 * 2. The response has not been dispatched to a callback.
+	 * 3. The response has not already been retrieved previously.
+	 */
+	std::optional<Response> get_response();
+};
+```
+
+#### Request Status
+
+```c++
+enum RequestStatus {
+	REQUEST_SUCCESS = 0,
+	REQUEST_ERROR = -1,
+	REQUEST_IN_PROGRESS = 1,
+};
+```
+
+#### Request Fan-Out
+
+```c++
+/**
+ * A fan-out encapsulates the collection of responses for a collection of 
+ * requests.
+ */
+class FanOut {
+	template<class InputIt>
+	FanOut(InputIt first, InputIt last)
+	FanOut(std::initializer_list<Request> requests);
+	
+	/**
+	 * Return a list of ready request, and cancel the requests for which a 
+	 * response is not available. 
+	 */
+	tnt::List<Request> collect();
+};
+```
```

Comment:
> What's about `waitFor`, we could just check deadlines for all combinators on every callback from `IOEventProvider`.

I believe we discussed that we don't want to implement timers, because the user has more precise time information.

> At this moment, we could fire combinator `any`, associated with the request, or bump and check counter for combinator `all`.

What interface do you propose for these combinators? It seems like this will create a lot of additional overhead (i.e., the connection objects will have to manage the combinator objects).